data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaSinkProducer.java

```diff
@@ -0,0 +1,171 @@
+import org.opensearch.dataprepper.plugins.kafka.configuration.TopicConfig;
+import org.opensearch.dataprepper.plugins.kafka.sink.DLQSink;
+import org.opensearch.dataprepper.plugins.kafka.util.MessageFormat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.LinkedList;
+
+
+/**
+ * * A helper class which helps takes the buffer data
+ * and produce it to a given kafka topic
+ */
+
+public class KafkaSinkProducer<T> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaSinkProducer.class);
+
+    private final Producer<String, T> producer;
+
+    private final KafkaSinkConfig kafkaSinkConfig;
+
+    private final DLQSink dlqSink;
+
+    private final CachedSchemaRegistryClient schemaRegistryClient;
+
+    private final Collection<EventHandle> bufferedEventHandles;
+
+    final ExpressionEvaluator expressionEvaluator;
+
+    public KafkaSinkProducer(final Producer producer,
+                             final KafkaSinkConfig kafkaSinkConfig,
+                             final DLQSink dlqSink,
+                             final CachedSchemaRegistryClient schemaRegistryClient,
+                             final ExpressionEvaluator expressionEvaluator) {
+        this.producer = producer;
+        this.kafkaSinkConfig = kafkaSinkConfig;
+        this.dlqSink = dlqSink;
+        this.schemaRegistryClient = schemaRegistryClient;
+        bufferedEventHandles = new LinkedList<>();
+        this.expressionEvaluator = expressionEvaluator;
+
+    }
+
+    public void produceRecords(final Record<Event> record) {
+        if (record.getData().getEventHandle() != null) {
+            bufferedEventHandles.add(record.getData().getEventHandle());
+        }
+        TopicConfig topic = kafkaSinkConfig.getTopic();
+        final Event event = record.getData();
+        final String key = event.formatString(kafkaSinkConfig.getPartitionKey(), expressionEvaluator);
+        Object dataForDlq = event.toJsonString();
+        LOG.info("Producing record " + dataForDlq);
+        try {
+            final String serdeFormat = kafkaSinkConfig.getSerdeFormat();
+            if (MessageFormat.JSON.toString().equalsIgnoreCase(serdeFormat)) {
+                publishJsonMessage(record, topic, key, dataForDlq);
+            } else if (MessageFormat.AVRO.toString().equalsIgnoreCase(serdeFormat)) {
+                publishAvroMessage(record, topic, key, dataForDlq);
+            } else {
+                publishPlaintextMessage(record, topic, key, dataForDlq);
+            }
+        } catch (Exception e) {
+            releaseEventHandles(false);
+        }
+
+
+    }
+
+    private void publishPlaintextMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) {
+        producer.send(new ProducerRecord(topic.getName(), key, record.getData().toJsonString()), callBack(dataForDlq));
```

Comment:
I just realized that you are not handling "tagsTargetKey". I think Umair/Omkar is familiar with this. You can also see OpenSearchSink code for example.

Reply:
Yes , we have incorporated the changes.